### PR TITLE
Narrow string on strlen() == and === comparison with integer-range

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11548

----

after implementing I realized this also makes `count($arr) == integer-range` narrow to `non-empty-array`

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -2172,6 +2172,29 @@ public function resolveIdentical(Expr\BinaryOp\Identical $expr, Scope $scope, Ty
 			}
 		}
 
+		if ($rightType instanceof IntegerRangeType) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php]: inspired by the previous loop with constant-strings, I have added a similar case for integer ranges

File: src/Analyser/TypeSpecifier.php [hunk 2]
```diff
@@ -1090,10 +1096,6 @@ private function specifyTypesForConstantBinaryExpression(
 			));
 		}
 
-		if ($constantType->getValue() === null) {
-			return $this->create($exprNode, $constantType, $context, false, $scope, $rootExpr);
-		}
-
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1096]: looked like dead code

File: src/Analyser/TypeSpecifier.php [hunk 3]
```diff
@@ -2149,10 +2157,19 @@ public function resolveIdentical(Expr\BinaryOp\Identical $expr, Scope $scope, Ty
 			}
 		}
 
-		if (count($rightType->getConstantStrings()) > 0) {
+		if ($rightType->isInteger()->yes() || $rightType->isString()->yes()) {
 			$types = null;
-			foreach ($rightType->getConstantStrings() as $constantString) {
-				$specifiedType = $this->specifyTypesForConstantStringBinaryExpression($unwrappedLeftExpr, $constantString, $context, $scope, $rootExpr);
+			foreach ($rightType->getFiniteTypes() as $finiteType) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:2170]: I need to check whether this leads to dead code in IntegerRange handling somewhere

File: src/Analyser/TypeSpecifier.php [hunk 4]
```diff
@@ -1102,6 +1104,10 @@ private function specifyTypesForConstantBinaryExpression(
 			&& in_array(strtolower((string) $exprNode->name), ['count', 'sizeof'], true)
 			&& $constantType instanceof ConstantIntegerType
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1105]: in a followup PR I would factor out the 2 `$constantType instanceof ConstantIntegerType` if-blocks of this method into the parent caller and do a proper integer range check instead, which will fix https://github.com/phpstan/phpstan/issues/11558

I don't do it now to keep this PR simpler and the next one will also be easier to review

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**